src/librbd/image/RemoveRequest.cc

```diff
@@ -0,0 +1,777 @@
+    }
+    return;
+  }
+  for (auto &watcher : mirror_watchers) {
+    m_watchers.remove_if([watcher] (obj_watch_t &w) {
+        return (strncmp(w.addr, watcher.addr, sizeof(w.addr)) == 0);
+      });
+  }
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_check_image_watchers(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error listing image watchers: " << cpp_strerror(*result) << dendl;
+    send_close_image(*result);
+    return nullptr;
+  }
+  // If an image is being bootstrapped by rbd-mirror, it implies
+  // that the rbd-mirror daemon currently has the image open.
+  // Permit removal if this is the case.
+  filter_out_mirror_watchers();
+
+  if (m_watchers.size() > 1) {
+    lderr(m_cct) << "image has watchers - not removing" << dendl;
+    send_close_image(-EBUSY);
+    return nullptr;
+  }
+
+  check_image_consistency_group();
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::check_image_consistency_group() {
+  ldout(m_cct, 20) << dendl;
+
+  librados::ObjectReadOperation op;
+  librbd::cls_client::image_get_group_start(&op);
+
+  using klass = RemoveRequest<I>;
+  librados::AioCompletion *rados_completion =
+    create_rados_ack_callback<klass, &klass::handle_check_image_consistency_group>(this);
+  m_out_bl.clear();
+  int r = m_image_ctx->md_ctx.aio_operate(m_header_oid,
+                                          rados_completion, &op, &m_out_bl);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_check_image_consistency_group(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error fetching consistency group for image" << cpp_strerror(*result)
+                 << dendl;
+    send_close_image(*result);
+    return nullptr;
+  }
+
+  cls::rbd::GroupSpec s;
+  bufferlist::iterator it = m_out_bl.begin();
+  *result = librbd::cls_client::image_get_group_finish(&it, &s);
+  if (*result < 0) {
+    send_close_image(*result);
+    return nullptr;
+  }
+  if (s.is_valid()) {
+    lderr(m_cct) << "image is in a consistency group - not removing" << dendl;
+    send_close_image(-EMLINK);
+    return nullptr;
+  }
+
+  trim_image();
+
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::trim_image() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = RemoveRequest<I>;
+  Context *ctx = create_async_context_callback(
+    *m_image_ctx, create_context_callback<klass, &klass::handle_trim_image>(this));
+
+  RWLock::RLocker owner_lock(m_image_ctx->owner_lock);
+  librbd::operation::TrimRequest<I> *req = librbd::operation::TrimRequest<I>::create(
+    *m_image_ctx, ctx, m_image_ctx->size, 0, m_prog_ctx);
+  req->send();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_trim_image(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  {
+    // Wait the RLocker released in trim_image()
```

Comment:
I am not sure why this is even needed. ```librbd::image::CloseRequest``` will eventually take a write-lock on ```ImageCtx::owner_lock```, so any potential for thread races should be cleared at that point long before the lock is destroyed. Did you ever hit a race where something failed because the trim request process was still holding a read-lock?